single-pool/js/packages/client/src/transactions.ts

```diff
@@ -0,0 +1,323 @@
+import { SINGLE_POOL_PROGRAM_ID } from './internal';
+import {
+  STAKE_PROGRAM_ID,
+  STAKE_ACCOUNT_SIZE,
+  MINT_SIZE,
+  StakeInstruction,
+  SystemInstruction,
+  TokenInstruction,
+  StakeAuthorizationType,
+  getAssociatedTokenAddress,
+} from './quarantine';
+
+interface DepositParams {
+  rpc: any; // XXX Rpc<???>
+  pool: PoolAddress;
+  userWallet: Base58EncodedAddress;
+  userStakeAccount?: Base58EncodedAddress;
+  depositFromDefaultAccount?: boolean;
+  userTokenAccount?: Base58EncodedAddress;
+  userLamportAccount?: Base58EncodedAddress;
+  userWithdrawAuthority?: Base58EncodedAddress;
+}
+
+interface WithdrawParams {
+  rpc: any; // XXX Rpc<???>
+  pool: PoolAddress;
+  userWallet: Base58EncodedAddress;
+  userStakeAccount: Base58EncodedAddress;
+  tokenAmount: bigint;
+  createStakeAccount?: boolean;
+  userStakeAuthority?: Base58EncodedAddress;
+  userTokenAccount?: Base58EncodedAddress;
+  userTokenAuthority?: Base58EncodedAddress;
+}
+
+export class SinglePoolProgram {
+  static programAddress: Base58EncodedAddress = SINGLE_POOL_PROGRAM_ID;
+  static space: bigint = 33n;
+
+  static async initialize(
+    rpc: any, // XXX not exported: Rpc<???>,
+    voteAccount: VoteAccountAddress,
+    payer: Base58EncodedAddress,
+    skipMetadata = false,
+  ): Promise<Transaction> {
+    let transaction = { instructions: [] as any, version: 'legacy' as TransactionVersion };
+
+    const pool = await findPoolAddress(this.programAddress, voteAccount);
+    const stake = await findPoolStakeAddress(this.programAddress, pool);
+    const mint = await findPoolMintAddress(this.programAddress, pool);
+
+    const poolRent = await rpc.getMinimumBalanceForRentExemption(this.space).send();
+    const stakeRent = await rpc.getMinimumBalanceForRentExemption(STAKE_ACCOUNT_SIZE).send();
+    const mintRent = await rpc.getMinimumBalanceForRentExemption(MINT_SIZE).send();
+    const minimumDelegation = (await rpc.getStakeMinimumDelegation().send()).value;
+
+    transaction = appendTransactionInstruction(
+      SystemInstruction.transfer({
+        from: payer,
+        to: pool,
+        lamports: poolRent,
+      }),
+      transaction,
+    );
+
+    transaction = appendTransactionInstruction(
+      SystemInstruction.transfer({
+        from: payer,
+        to: stake,
+        lamports: stakeRent + minimumDelegation,
+      }),
+      transaction,
+    );
+
+    transaction = appendTransactionInstruction(
+      SystemInstruction.transfer({
+        from: payer,
+        to: mint,
+        lamports: mintRent,
+      }),
+      transaction,
+    );
+
+    transaction = appendTransactionInstruction(
+      await SinglePoolInstruction.initializePool(voteAccount),
+      transaction,
+    );
+
+    if (!skipMetadata) {
+      transaction = appendTransactionInstruction(
+        await SinglePoolInstruction.createTokenMetadata(pool, payer),
+        transaction,
+      );
+    }
+
+    return transaction;
```

Comment:
We gotchu in `@solana/functional` now.

```ts
return pipe(
  createTransaction({ version: 'legacy' }),
  tx => appendTransactionInstruction(/* ... */, tx),
  tx => appendTransactionInstruction(/* ... */, tx),
  tx => appendTransactionInstruction(/* ... */, tx),
  tx => appendTransactionInstruction(/* ... */, tx),
  tx => skipMetadata ? tx : appendTransactionInstruction(/* ... */, tx),
);
```

cc/ @buffalojoec 🎉

Reply:
Not 100% sure how the `await`s will perform in the `pipe` but you might get away with `Promise.all()`!

Reply:
You _have_ to resolve all the awaitables outside of the pipe using `Promise.all`, yeah.